intr_handler.c

```diff
@@ -0,0 +1,45 @@
+#include <intr_handler.h>
+
+#include <common.h>
+#include <queue.h>
+
+#include "libkern.h"
+#include <pic32mz.h>
+
+intr_event_t *events[PIC32_IRQ_LAST];
+
+void intr_event_add_handler(intr_event_t *ie, intr_handler_t *ih) {
+  intr_handler_t *it;
+  /* Add new handler according to it's priority */
+  TAILQ_FOREACH(it, &ie->ie_handlers, ih_next) {
+    if (ih->ih_prio > it->ih_prio)
+      break;
+  }
+  if (it)
+    TAILQ_INSERT_BEFORE(it, ih, ih_next);
+  else
+    /* List is empty */
+    TAILQ_INSERT_TAIL(&ie->ie_handlers, ih, ih_next);
+}
+
+void run_event_handlers(unsigned irq) {
+  intr_event_execute_handlers(events[irq]);
+}
+
+void intr_event_init(intr_event_t *ie, uint8_t irq, char *name) {
+  ie->ie_rq = irq;
+  ie->ie_name = name;
+  TAILQ_INIT(&ie->ie_handlers);
+  events[irq] = ie;
+}
+
+void intr_event_remove_handler(intr_handler_t *ih) {
+  TAILQ_REMOVE(&ih->ih_event->ie_handlers, ih, ih_next);
+}
+
+void intr_event_execute_handlers(intr_event_t *ie) {
+  intr_handler_t *it;
+  TAILQ_FOREACH(it, &ie->ie_handlers, ih_next) {
```

Comment:
What will happen if all handler report FILTER_STRAY? Will a spurious interrupt request be handled correctly, so that it will not be triggered again just after interrupt handler routine finishes?